nuxeo-runtime/nuxeo-connect-standalone/src/test/java/org/nuxeo/launcher/connect/TestConnectBroker.java

```diff
@@ -284,12 +285,66 @@ public void testDownloadUnknownPackage() throws Exception {
 
         // Assertions
         checkPackagesState(connectBroker, Arrays.asList("unknown-package"), NON_EXISTENT);
+        connectBroker.getCommandSet().log();
         List<String> expectedLogs = new ArrayList<>();
         expectedLogs.add("Downloading [unknown-package]...");
         expectedLogs.add("\tDownload failed (not found).");
         checkLogEvents(expectedLogs, logCaptureResult.getCaughtEvents());
     }
 
+    /**
+     * Reproduction of the commands in <a href="https://jira.nuxeo.com/browse/NXP-21573">NXP-21573</a>, which should be
+     * successful:
+     *
+     * <pre><code>
+     * mp-add $TEST_LOCAL_ONLY_PATH/F-1.0.0-SNAPSHOT.zip
+     * mp-install F
+     * </code></pre>
+     */
+    @Test
+    @LogCaptureFeature.FilterWith(PkgRequestLogFilter.class)
+    public void testDownloadUnknownSnapshotPackage() throws Exception {
+        // Environment setup
+        Environment.getDefault().setProperty(Environment.DISTRIBUTION_NAME, "server");
+        Environment.getDefault().setProperty(Environment.DISTRIBUTION_VERSION, "8.3");
+        ConnectBroker connectBroker = new ConnectBroker(Environment.getDefault());
+        ((StandaloneCallbackHolder) NuxeoConnectClient.getCallBackHolder()).setTestMode(true);
+        connectBroker.setAllowSNAPSHOT(true);
+        final PackageState NON_EXISTENT = null;
+
+        // Given an unknown package 'F-1.0.0-SNAPSHOT', which is:
+        // - only available from local file system
+        // - not yet added into the Nuxeo Server 8.3
+        assertThat(new File(TEST_LOCAL_ONLY_PATH + "/F-1.0.0-SNAPSHOT.zip")).exists();
+        checkPackagesState(connectBroker, Arrays.asList("F-1.0.0-SNAPSHOT"), NON_EXISTENT);
+
+        // When enable the relax mode, and mp-add $TEST_LOCAL_ONLY_PATH/F-1.0.0-SNAPSHOT.zip
+        connectBroker.setRelax("true");
+        connectBroker.pkgAdd(TEST_LOCAL_ONLY_PATH + "/F-1.0.0-SNAPSHOT.zip", false);
+
+        // Then the package is added
+        List<String> startedPackages = new ArrayList<>();
+        startedPackages.add("F-1.0.0-SNAPSHOT");
+        checkPackagesState(connectBroker, startedPackages, PackageState.DOWNLOADED);
+
+        // When mp-install F
+        boolean isSuccessful = connectBroker.pkgRequest(null, Arrays.asList("F"), null, null, true, true);
+
+        // Then it's successful
+        assertThat(isSuccessful).isTrue();
+        connectBroker.getCommandSet().log();
+        String expectedLog = "Added src/test/resources/packages/store/local-only/F-1.0.0-SNAPSHOT.zip\n" + //
```

Comment:
Yes, method `logOf(...)` is an improvement compared to the previous method `checkLogEvents(...)` because:

- The previous method `checkLogEvents(...)` asserts the number of logs, before asserting the log content: so when the number of lines changes, the message will be like _"Expecting the nb of logs to be 5, but got 7..."_ without any detail. This is very confusing.
- The current method `logOf(...)` ensures that log content can be asserted directly.

This method is `static` because the input-argument `LogCaptureFeature.Result` is the only data manipulated by this utility method. In a static context, this method gives up the right to access any non-static / instance member of the class, so **it ensures that no instance member can be modified by this method, except the method input**. Any attempt of adding an instance member into the static method will fail at compilation. (I think it's overkill here... If it's too confusing, I can delete it.)

All the usages can be and should be replaced by method `logOf(...)`. I didn't do it previously because I want to provide a clearer `git-diff` to facilitate the code review 🙂